processors/enduserprocessor/enduserprocessor.go

```diff
@@ -0,0 +1,535 @@
+	enduserSessionAttribute = "session.id"
+)
+
+type config struct {
+	EndUser
+	hashAlgorithm hash.Algorithm
+}
+
+type processor struct {
+	logger             *zap.Logger
+	attributeConfigMap map[string][]config
+}
+
+type user struct {
+	id      string
+	role    string
+	scope   string
+	session string
+}
+
+var _ processorhelper.TProcessor = (*processor)(nil)
+
+func (p *processor) ProcessTraces(_ context.Context, traces pdata.Traces) (pdata.Traces, error) {
+	rss := traces.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+
+		ilss := rs.InstrumentationLibrarySpans()
+		for j := 0; j < ilss.Len(); j++ {
+			ils := ilss.At(j)
+
+			spans := ils.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				span.Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+					unindexedKey := piifilterprocessor.UnindexedKey(k)
+					if configs, ok := p.attributeConfigMap[unindexedKey]; ok {
+						for _, cfg := range configs {
+							p.capture(span, cfg, v.StringVal())
+						}
+					}
+				})
+			}
+		}
+	}
+
+	return traces, nil
+}
+
+func (p *processor) capture(span pdata.Span, enduser config, value string) {
+	if !p.passesConditions(span, enduser.Conditions) {
+		return
+	}
+
+	var user *user
+	switch enduser.Type {
+	case "id":
+		user = p.idCapture(value)
+	case "role":
+		user = p.roleCapture(value)
+	case "scope":
+		user = p.scopeCapture(value)
+	case "authheader":
+		user = p.authHeaderCapture(enduser, value)
+	case "json":
+		user = p.jsonCapture(enduser, value)
+	case "urlencoded":
+		user = p.urlencodedCapture(enduser, value)
+	case "cookie":
+		user = p.cookieCapture(enduser, value)
+	default:
+		p.logger.Warn("Unknown enduser type, skipping", zap.String("type", enduser.Type), zap.String("key", enduser.AttributeKey))
+	}
+
+	if user == nil {
+		return
+	}
+
+	if len(user.id) > 0 {
+		addSpanAttribute(span, enduserIDAttribute, user.id)
+	}
+
+	if len(user.role) > 0 {
+		addSpanAttribute(span, enduserRoleAttribute, user.role)
+	}
+
+	if len(user.scope) > 0 {
+		addSpanAttribute(span, enduserScopeAttribute, user.scope)
+	}
+
+	if len(user.session) > 0 {
+		addSpanAttribute(span, enduserSessionAttribute, user.session)
+	}
+}
+
+func addSpanAttribute(span pdata.Span, key string, value string) {
+	attributes := span.Attributes()
+	// don't overwrite existing attributes
+	if _, ok := attributes.Get(key); ok {
+		return
+	}
+	attributes.Insert(key, pdata.NewAttributeValueString(value))
+}
+
+// only capture the user info if all conditions
+// are true.  If an conditions key does not exist
+// in the span, that is considered a failed condition
+func (p *processor) passesConditions(span pdata.Span, conditions []Condition) bool {
+	attribMap := span.Attributes()
+
+	for _, condition := range conditions {
+		attrib, ok := attribMap.Get(condition.Key)
+		if !ok {
+			return false
+		}
+
+		value := attrib.StringVal()
+		// TODO is it correct? it was `value == nil` before
```

Comment:
@davexroth can you please check this?

Reply:
Yep, this is correct.  In the oc collector GetStringValue() returned a pointer.